test/refactoring/ConvertAsync/convert_bool.swift

```diff
@@ -0,0 +1,448 @@
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path | %FileCheck -check-prefix=BOOL-WITH-ERR-SILLY2 %s
+boolWithErr { success, err in
+  if success {
+    print("ummm wat \(err!)")
+  } else {
+    print("ummm wat \(err!)")
+  }
+}
+
+// The err unwrap is in both blocks, so it's not clear what to classify as.
+
+// BOOL-WITH-ERR-SILLY2:      let success = try await boolWithErr()
+// BOOL-WITH-ERR-SILLY2-NEXT: if success {
+// BOOL-WITH-ERR-SILLY2-NEXT:   print("ummm wat \(<#err#>!)")
+// BOOL-WITH-ERR-SILLY2-NEXT: } else {
+// BOOL-WITH-ERR-SILLY2-NEXT:   print("ummm wat \(<#err#>!)")
+// BOOL-WITH-ERR-SILLY2-NEXT: }
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path | %FileCheck -check-prefix=MULTI-BOOL-WITH-ERR %s
+multipleBoolWithErr { str, b1, b2, err in
+  if !b1 && !b2 {
+    print("a \(err!)")
+  }
+  if b1, b2 {
+    print("b \(err!)")
+  }
+  if !b1 {
+    print("c \(err!)")
+  }
+  if !b2 {
+    print("d \(err!)")
+  }
+}
+
+// Don't handle the case where multiple flag checks are done in a single
+// condition, because it's not exactly clear what the user is doing. It's a
+// little unfortunate that we'll allow multiple flag checks in seperate
+// conditions, but both of these cases seem somewhat uncommon, and there's no
+// real way to completely enforce a single flag param across e.g multiple calls
+// to the same function, so this is probably okay for now.
+
+// MULTI-BOOL-WITH-ERR:      do {
+// MULTI-BOOL-WITH-ERR-NEXT:   let (str, b1, b2) = try await multipleBoolWithErr()
+// MULTI-BOOL-WITH-ERR-NEXT: } catch let err {
+// MULTI-BOOL-WITH-ERR-NEXT:   if !<#b1#> && !<#b2#> {
+// MULTI-BOOL-WITH-ERR-NEXT:     print("a \(err)")
+// MULTI-BOOL-WITH-ERR-NEXT:   }
+// MULTI-BOOL-WITH-ERR-NEXT:   if <#b1#>, <#b2#> {
+// MULTI-BOOL-WITH-ERR-NEXT:     print("b \(err)")
+// MULTI-BOOL-WITH-ERR-NEXT:   }
+// MULTI-BOOL-WITH-ERR-NEXT:   print("c \(err)")
+// MULTI-BOOL-WITH-ERR-NEXT:   print("d \(err)")
+// MULTI-BOOL-WITH-ERR-NEXT: }
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path | %FileCheck -check-prefix=OPT-BOOL-WITH-ERR %s
+optionalBoolWithErr { str, optBool, b, err in
+  if optBool != nil {
+    print("a \(err!)")
+  }
+  if optBool == nil {
+    print("b \(err!)")
+  }
+  if optBool == true {
+    print("c \(err!)")
+  }
+  if ((optBool) == (false)) {
+    print("d \(err!)")
+  }
+  if optBool == false {
+    print("e \(err)")
+  }
+  if optBool != true {
+    print("f \(err!)")
+  }
+  if b {
+    print("g \(err!)")
+  }
+}
+
+// It's a little unfortunate that print("a \(<#err#>!)") gets classified in the success
+// block below, but it doesn't seem like a case that's likely to come up, as optBool
+// would then be inaccessible in the error block.
```

Comment:
For both `a` and `f` we could just assume that if there's a force unwrap, that == the error branch (only where the conditions aren't explicitly success/error and involve known params). I'm happy with this as is though, just thought I'd mention it.

Reply:
Yeah it's possible we'll want to generalise the heuristic to better handle cases like these, though I'm not really sure how often they'll come up in practice as it seems a little odd to e.g check a success param for non-nil and then assume you're on an error path without looking at the success param again. I think for now at least I'd rather tend on the side on keeping the heuristic a little more limited and seeing what kind of feedback we get from users.